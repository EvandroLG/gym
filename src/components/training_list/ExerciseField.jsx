import React from 'react';

export default function ExerciseField(props) {
  return (
    <td>
      <input
        type="text"
        onChange={ (e) => props.onInputChange(e, props.index, props.property) }
        value={ props.value }
      />
    </td>
  ) 
}
