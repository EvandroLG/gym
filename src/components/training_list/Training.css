.list-group h3 {
  margin: 20px 0;
  font-size: 20px;
}

.list-group table {
  width: 100%;
  border-collapse: collapse;
}

.list-group tr {
  border-collapse: separate;
}

.list-group thead {
  background: #222;
  color: #FFF;
}

.list-group td, .list-group th {
  padding: 7px;
}

.list-group tbody tr:hover {
  background: #F6F7F9; 
}

.list-group button.edit {
  margin-right: 5px;
}

.list-group form {
  margin-top: 20px;
  position: relative;
}

.list-group form > input[type="text"] {
  width: 100%;
  margin-bottom: 10px;
}

.list-group button.add, .list-group button.remove {
  float: right;
}

.list-group li {
  margin-bottom: 30px;
}

.list-group li:last-child {
  margin-bottom: 0;
}
