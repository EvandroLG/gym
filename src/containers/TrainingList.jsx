import { connect } from 'react-redux';
import DB from '../libs/db';
import TrainingList from '../components/training_list/';

const mapStateToProps = ({ trainingList }) => {
  return {
    trainingList
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    findAll: () => {
      new DB().findAll((trainingList) => {
        dispatch({
          type: "LOAD_TRAINING_LIST",
          trainingList,
        });
      });
    },

    onRemoveTraining: (id) => {
      new DB().remove(id);

      dispatch({
        type: 'REMOVE_TRAINING',
        id
      });
    },

    onAddNewExercise: (id) => {
      dispatch({
        type: 'ADD_NEW_EXERCISE',
        id
      });
    },

    onRemoveExercise: (idTraining, idExercise) => {
      dispatch({
        type: 'REMOVE_EXERCISE',
        idTraining,
        idExercise
      });
    },

    onChangeTitle: (id, title) => {
      dispatch({
        type: 'CHANGE_TITLE_TRAINING',
        id,
        title
      });
    },

    onChangeExercise: (idTraining, idExercise, property, value) => {
      dispatch({
        type: 'CHANGE_EXERCISE_TRAINING',
        idTraining,
        idExercise,
        property,
        value
      });
    },

    onSubmit: (id, training) => {
      new DB().update(id, training);
    }
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(TrainingList);
