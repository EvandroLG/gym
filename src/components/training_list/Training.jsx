import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setVideoUrl } from '../../action_creators';
import ExerciseField from './ExerciseField';
import DB from '../../libs/db';

class Training extends Component {

  constructor(props) {
    super(props);

    this.db = new DB();
    this.state = {
      editing: false,
      title: this.props.title,
      exercises: this.props.exercises
    };
  }

  _renderButtonWatchVideo(url) {
    if (url) {
      return (
        <td>
          <button type="button" onClick={ () => this.props.onVideoUrlChange(url) }>
            Watch
          </button>

        </td>
      );
    }

    return <td>---</td>;
  }

  _renderExerciseTr(exercise, key) {
    return (
      <tr key={ key }>
        <td>{ exercise.name }</td>
        <td>{ exercise.set }</td>
        <td>{ exercise.repetition }</td>
        <td>{ exercise.weight }</td>
        { this._renderButtonWatchVideo(exercise.youtube) }
        <td></td>
      </tr>
    )
  }

  _onExerciseInputChange(e, key, property) {
    let exercises = this.state.exercises;
    exercises[key][property] = e.target.value;

    this.setState({ exercises });
  }

  _onRemoveButtonClick(index) {
    this.state.exercises.splice(index, 1);
    this.setState({
      exercises: this.state.exercises
    });
  }

  _renderExerciseForm(exercise, key) {
    const onExerciseInputChange = this._onExerciseInputChange.bind(this);

    return (
      <tr key={ key }>
        <ExerciseField value={ exercise.name } index={ key } property='name'
          onInputChange={ onExerciseInputChange } />
        <ExerciseField value={ exercise.set } index={ key } property='set'
          onInputChange={ onExerciseInputChange } />
        <ExerciseField value={ exercise.repetition } index={ key } property='repetition'
          onInputChange={ onExerciseInputChange } />
        <ExerciseField value={ exercise.weight } index={ key } property='weight'
          onInputChange={ onExerciseInputChange } />
        <ExerciseField value={ exercise.youtube } index={ key } property='youtube'
          onInputChange={ onExerciseInputChange } />

        <td>
          <button type="button" className="btn btn-danger btn-sm"
            onClick={ () => this._onRemoveButtonClick(key) }>
            Remove
          </button>
        </td>
      </tr>
    )
  }

  _renderExerciseList() {
    const editing = this.state.editing;

    return this.state.exercises.map((exercise, key) => {
      return this[editing ? '_renderExerciseForm' : '_renderExerciseTr'](exercise, key);
    });
  }

  _onEditButtonClick() {
    this.setState({
      editing: true
    });
  }

  _onInputTitleChange(e) {
    this.setState({
      title: e.target.value
    });
  }

  _renderInputTitle() {
    return (
      <input type="text" onChange={ this._onInputTitleChange.bind(this) }
        value={ this.state.title } />
    )
  }

  _renderTitle() {
    return (
      <h4>{ this.state.title }</h4>
    )
  }

  _onButtonAddClick() {
    this.setState({
      exercises: this.state.exercises.concat({
        name: '',
        set: '',
        repetition: '',
        weight: '',
        youtube: ''
      })
    });
  }

  _renderAddButton() {
    if (!this.state.editing) { return; }

    return (
      <tr>
        <td colSpan="5">
          <button type="button" onClick={ this._onButtonAddClick.bind(this) }
            className="btn btn-sm btn-primary float-right">
            Add
          </button>
        </td>
      </tr>
    )
  }

  _renderTable() {
    return (
      <table className="table table-hover">
        <thead className="thead-dark">
          <tr>
            <th className="col">Exercise</th>
            <th className="col">Set</th>
            <th className="col">Repetition</th>
            <th className="col">Weight</th>
            <th className="col">Video</th>
            <th className="col"></th>
          </tr>
        </thead>
        <tbody>
          { this._renderExerciseList() }
          { this._renderAddButton() }
        </tbody>
      </table>
    )
  }

  _onFormSubmit(e) {
    e.preventDefault();

    const exercises = this.state.exercises.filter((exercise) => {
      return exercise.name;
    });

    const props = {
      title: this.state.title,
      exercises
    };

    this.setState({ exercises });

    this.db.update(this.props.index, props, () => {
      this.setState({
        editing: false
      });
    });
  }

  _renderSubmitButton() {
    if (!this.state.editing) { return; }

    return (
      <input type="submit" className="btn btn-success btn-sm float-right" value="Save!" />
    )
  }

  render() {
    return (
      <li className="list-group-item">
        <div className="btn-group float-right mb-1">
          <button type="button" className="btn btn-primary btn-sm"
            onClick={ this._onEditButtonClick.bind(this) }>
            Edit
          </button>

          <button type="button" className="btn btn-danger btn-sm"
            onClick={ () => this.props.onButtonClick(this.props.index) }>
            Remove
          </button>
        </div>

        <form onSubmit={ this._onFormSubmit.bind(this) }>
          { this[this.state.editing ? '_renderInputTitle' : '_renderTitle']() }
          { this._renderTable() }
          { this._renderSubmitButton() }
        </form>
      </li>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    videoUrl: state.videoUrl
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    onVideoUrlChange: (url) => {
      dispatch(setVideoUrl(url));
    }
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Training);
