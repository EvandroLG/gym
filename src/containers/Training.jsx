import { connect } from 'react-redux';
import { setVideoUrl } from '../action';
import Training from '../components/training_list/Training'

const mapStateToProps = ({ videoUrl }) => {
  return {
    videoUrl
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    onVideoUrlChange: (url) => {
      dispatch(setVideoUrl(url));
    }
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Training);
