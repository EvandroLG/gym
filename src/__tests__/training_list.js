import React from 'react';
import { shallow } from 'enzyme';
import DB from '../libs/db';
import ExerciseField from '../components/training_list/exercise_field';
import Training from '../components/training_list/training';
import TrainingList from '../components/training_list/index';
import fixtures from './fixtures';

describe('training list', () => {
  describe('instances', () => {
    it('execise field component should render as expected', () => {
      const component = shallow(<ExerciseField />);
      verifySnapshot(component);
    });

    it('training component should render as expected', () => {
      DB.prototype._initialize = () => {}; 
      DB.prototype.update = () => {};

      const props = {
        index: '1',
        title: fixtures[0].title,
        exercises: fixtures[0].exercises,
        onButtonClick: () => {}
      };

      const component = shallow(<Training key={ props.index } { ...props } />);
      verifySnapshot(component);
    });

    it('training list component should render as expected', () => {
      DB.prototype.findAll = (callback) => {
        callback(fixtures);
      };

      const component = shallow(<TrainingList />);
      verifySnapshot(component);
    });
  });
});
