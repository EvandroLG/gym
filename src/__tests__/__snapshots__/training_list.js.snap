// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`training list instances execise field component should render as expected 1`] = `
<td>
  <input
    onChange={[Function]}
    type="text"
  />
</td>
`;

exports[`training list instances training component should render as expected 1`] = `
<li
  className="list-group-item"
>
  <div
    className="btn-group float-right mb-1"
  >
    <button
      className="btn btn-primary btn-sm"
      onClick={[Function]}
      type="button"
    >
      Edit
    </button>
    <button
      className="btn btn-danger btn-sm"
      onClick={[Function]}
      type="button"
    >
      Remove
    </button>
  </div>
  <form
    onSubmit={[Function]}
  >
    <h4>
      Leg
    </h4>
    <table
      className="table table-hover"
    >
      <thead
        className="thead-dark"
      >
        <tr>
          <th
            className="col"
          >
            Exercise
          </th>
          <th
            className="col"
          >
            Set
          </th>
          <th
            className="col"
          >
            Repetition
          </th>
          <th
            className="col"
          >
            Weight
          </th>
          <th
            className="col"
          />
        </tr>
      </thead>
      <tbody>
        <tr
          key="0"
        >
          <td>
            Barbell Deadlift
          </td>
          <td>
            3
          </td>
          <td>
            8
          </td>
          <td>
            10kg
          </td>
          <td />
        </tr>
        <tr
          key="1"
        >
          <td>
            Wide-Grip Pull-Up
          </td>
          <td>
            3
          </td>
          <td>
            8
          </td>
          <td>
            35kg
          </td>
          <td />
        </tr>
      </tbody>
    </table>
  </form>
</li>
`;

exports[`training list instances training list component should render as expected 1`] = `
<div>
  <h1>
    Current Training
  </h1>
  <ul
    className="list-group"
  >
    <Training
      exercises={
        Array [
          Object {
            "name": "Barbell Deadlift",
            "repetition": "8",
            "set": "3",
            "weight": "10kg",
          },
          Object {
            "name": "Wide-Grip Pull-Up",
            "repetition": "8",
            "set": "3",
            "weight": "35kg",
          },
        ]
      }
      index="1"
      key="1"
      onButtonClick={[Function]}
      title="Leg"
    />
    <Training
      exercises={
        Array [
          Object {
            "name": "Flat Bench Dumbbell Press",
            "repetition": "8",
            "set": "3",
            "weight": "22kg",
          },
          Object {
            "name": "Barbell Bench Press",
            "repetition": "8",
            "set": "3",
            "weight": "25kg",
          },
        ]
      }
      index="2"
      key="2"
      onButtonClick={[Function]}
      title="Chest"
    />
  </ul>
</div>
`;
